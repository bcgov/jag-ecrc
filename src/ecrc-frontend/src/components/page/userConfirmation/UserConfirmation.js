/* eslint-disable camelcase */
import React, { useState, useEffect } from "react";
import { Redirect } from "react-router-dom";
import PropTypes from "prop-types";
import axios from "axios";
import Header from "../../base/header/Header";
import Footer from "../../base/footer/Footer";
import { Button } from "../../base/button/Button";
import "./UserConfirmation.css";
import {
  generateJWTToken,
  accessJWTToken,
  isAuthenticated
} from "../../../modules/AuthenticationHelper";
import Loader from "../../base/loader/Loader";

export default function UserConfirmation({
  page: { header, setApplicant, setError }
}) {
  const [toConsent, setToConsent] = useState(false);
  const [toHome, setToHome] = React.useState(false);
  const [toTransition, setToTransition] = useState(false);
  const [toError, setToError] = useState(false);
  const [user, setUser] = useState({});
  const [fullName, setFullName] = useState("");
  const [toggleLoader, setToggleLoader] = useState({ display: "inline-block" });

  useEffect(() => {
    const urlParams = new URLSearchParams(window.location.search);
    const code = urlParams.get("code");

    if (!isAuthenticated() || !code) setToHome(true);

    const token = sessionStorage.getItem("jwt");
    const uuid = sessionStorage.getItem("uuid");

    axios
      .get(`/ecrc/protected/login?code=${code}&requestGuid=${uuid}`, {
        headers: {
          Authorization: `Bearer ${token}`
        }
      })
      .then(res => {
        sessionStorage.setItem("jwt", res.data);

        const {
          userInfo: {
            birthdate,
            address: { street_address, locality, region, postal_code },
            gender,
            given_name,
            given_names,
            family_name,
            identity_assurance_level
          }
        } = accessJWTToken(res.data);

        // TODO Check identity assurance level
        if (identity_assurance_level < 3) {
          setToTransition(true);
        }

        // Convert gender text
        const genderTxt = gender === "female" ? "F" : "M";

        // Convert date format
        const birthDt = birthdate.split("-").join("/");

        // Convert given names
        const givenNamesArray = given_names.split(" ");

        givenNamesArray.shift();

        const legalSecondNm = givenNamesArray.join(" ");

        // Convert province name
        const regionMap = new Map([
          ["BC", "BRITISH COLUMBIA"],
          ["AB", "ALBERTA"],
          ["NL", "NEWFOUNDLAND"],
          ["PE", "PRINCE EDWARD ISLAND"],
          ["NS", "NOVA SCOTIA"],
          ["NB", "NEW BRUNSWICK"],
          ["QC", "QUEBEC"],
          ["ON", "ONTARIO"],
          ["MB", "MANITOBA"],
          ["SK", "SASKATCHEWAN"],
          ["YT", "YUKON"],
          ["NT", "NORTH WEST TERRITORIES"],
          ["NU", "NUNAVUT"]
        ]);

        let provinceNm = regionMap.get(region);
        if (provinceNm === undefined) {
          provinceNm = "Invalid Province";
        }

        setUser({
          legalFirstNm: given_name,
          legalSecondNm,
          legalSurnameNm: family_name,
          birthDt,
          genderTxt,
          addressLine1: street_address,
          cityNm: locality,
          provinceNm,
          postalCodeTxt: postal_code,
          countryNm: "CANADA"
        });

        setFullName(`${given_name} ${family_name}`);
        setToggleLoader({ display: "none" });
      })
      .catch(error => {
        setToError(true);
        setError(error.response.status.toString());
        setToggleLoader({ display: "none" });
      });
    window.scrollTo(0, 0);
  }, [setError]);

  const yesButton = {
    label: "Yes",
    buttonStyle: "btn btn-primary",
    buttonSize: "btn",
    type: "submit"
  };

  const noButton = {
    label: "No",
    buttonStyle: "btn ecrc_accessary_btn",
    buttonSize: "btn",
    type: "submit"
  };

  function onYesClick() {
    const currentPayload = accessJWTToken(sessionStorage.getItem("jwt"));
    const newPayload = {
      ...currentPayload,
      actionsPerformed: ["userConfirmation"]
    };
    generateJWTToken(newPayload);

    setApplicant(user);
    setToConsent(true);
  }

  if (toConsent) {
    return <Redirect to="/criminalrecordcheck/userconfirmation" />;
  }

  if (toTransition) {
    return <Redirect to="/criminalrecordcheck/transition" />;
  }

  if (toHome) {
    return <Redirect to="/" />;
  }

  if (toError) {
    return <Redirect to="/criminalrecordcheck/error" />;
  }

  return (
    <main>
      <Header header={header} />
      <div className="page">
        <div className="user-confirm-content col-md-12">
          <strong>
            Please confirm the name associated with the BC Service card login
            provided.
          </strong>
          <p />
          <div style={toggleLoader}>
            <Loader page />
          </div>
          <p>{fullName}</p>
          <p>Is this correct?</p>
          <div className="row">
            <div className="col-md-12">
              <Button button={yesButton} onClick={() => onYesClick()} />
              <Button button={noButton} onClick={() => {}} />
              <br />
            </div>
          </div>
        </div>
      </div>
      <Footer />
    </main>
  );
}

UserConfirmation.propTypes = {
  page: PropTypes.shape({
    header: PropTypes.shape({
      name: PropTypes.string.isRequired
    }).isRequired,
    setApplicant: PropTypes.func.isRequired,
    setError: PropTypes.func.isRequired
  }).isRequired
};
