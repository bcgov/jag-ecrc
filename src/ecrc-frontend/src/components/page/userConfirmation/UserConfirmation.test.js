import React from "react";
import { create } from "react-test-renderer";
import { MemoryRouter } from "react-router-dom";

import UserConfirmation from "./UserConfirmation";

describe("User Confirmation Page Component", () => {
  test("Matches the snapshot", () => {
    const header = {
      name: "Criminal Record Check"
    };

    const setApplicant = () => {};

    const page = {
      header,
      setApplicant
    };

    const userConfirmation = create(
      <MemoryRouter>
        <UserConfirmation page={page} />
      </MemoryRouter>
    );
    expect(userConfirmation.toJSON()).toMatchSnapshot();
  });
});
