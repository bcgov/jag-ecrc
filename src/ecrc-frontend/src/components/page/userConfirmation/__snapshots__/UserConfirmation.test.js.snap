// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Confirmation Page Component Matches the snapshot 1`] = `
<main>
  <header>
    <nav
      className="navbar navbar-expand-lg navbar-dark"
    >
      <div
        className="container"
      >
        <a
          className="navbar-brand"
          href="/"
        >
          <img
            alt="B.C. Government Logo"
            className="img-fluid d-none d-md-block"
            height="44"
            src="/images/bcid-logo-rev-en.svg"
            width="181"
          />
          <img
            alt="B.C. Government Logo"
            className="img-fluid d-md-none"
            height="44"
            src="/images/bcid-symbol-rev.svg"
            width="64"
          />
        </a>
        <div>
          <a
            className="navbar-brand nav-item nav-link"
            href="/"
          >
            Criminal Record Check
          </a>
        </div>
        <button
          aria-controls="navbarNavAltMarkup"
          aria-expanded="false"
          aria-label="Toggle navigation"
          className="navbar-toggler"
          data-target="#navbarNavAltMarkup"
          data-toggle="collapse"
          type="button"
        >
          <span
            className="navbar-toggler-icon"
          />
        </button>
        <div
          className="collapse navbar-collapse"
          id="navbarNavAltMarkup"
        />
      </div>
    </nav>
  </header>
  <div
    className="page"
  >
    <div
      className="user-confirm-content col-md-12"
    >
      <strong>
        Please confirm the name associated with the BC Service card login provided.
      </strong>
      <p />
      <div
        style={
          Object {
            "display": "inline-block",
          }
        }
      >
        <div
          className="page-loader"
        />
      </div>
      <p>
        
      </p>
      <p>
        Is this correct?
      </p>
      <div
        className="row"
      >
        <div
          className="col-md-12"
        >
          <button
            className="btn btn-primary btn"
            onClick={[Function]}
            type="submit"
          >
            Yes
            <div
              className="loader-hide"
            >
              <div
                className="btn-loader"
              />
            </div>
          </button>
          <button
            className="btn ecrc_accessary_btn btn"
            onClick={[Function]}
            type="submit"
          >
            No
            <div
              className="loader-hide"
            >
              <div
                className="btn-loader"
              />
            </div>
          </button>
          <br />
        </div>
      </div>
    </div>
  </div>
  <footer
    className="footer"
  >
    <nav
      className="navbar navbar-expand-sm navbar-dark"
    >
      <div
        className="container"
      >
        <ul
          className="navbar-nav"
        >
          <li
            className="nav-item"
          >
            <a
              className="nav-link"
              href="https://www2.gov.bc.ca/gov/content/home/disclaimer"
              rel="noopener noreferrer"
              target="_blank"
            >
              Disclaimer
            </a>
          </li>
          <li
            className="nav-item"
          >
            <a
              className="nav-link"
              href="https://www2.gov.bc.ca/gov/content/home/privacy"
              rel="noopener noreferrer"
              target="_blank"
            >
              Privacy
            </a>
          </li>
          <li
            className="nav-item"
          >
            <a
              className="nav-link"
              href="https://www2.gov.bc.ca/gov/content/home/accessibility"
              rel="noopener noreferrer"
              target="_blank"
            >
              Accessibility
            </a>
          </li>
          <li
            className="nav-item"
          >
            <a
              className="nav-link"
              href="https://www2.gov.bc.ca/gov/content/home/copyright"
              rel="noopener noreferrer"
              target="_blank"
            >
              Copyright
            </a>
          </li>
          <li
            className="nav-item"
          >
            <a
              className="nav-link"
              href="https://www2.gov.bc.ca/gov/content/home/contact-us"
              rel="noopener noreferrer"
              target="_blank"
            >
              Contact Us
            </a>
          </li>
        </ul>
      </div>
    </nav>
  </footer>
</main>
`;
